### PR TITLE
[nrf52840] use nrf_security for hardware accelerated cryptography

### PR DESCRIPTION
**Pull request by KaSroka**
Previously simple CC310 cryptographic accelerator library was used with the glue layer using MbedTLS _ALT function substitution mechanism. This solution wasn't well tested from a security point of view.

The new cryptographic library is built from the nrf_security project: https://github.com/NordicPlayground/nrfxlib/tree/master/nrf_security. It provides both CC310 libraries with MbedTLS API and special glue layers for situations where both hardware and software implementation is used (i.e for AES as the hardware supports only 128bit keys so software is used for 256bit keys).

The new library undergoes special security verification procedure. The CC310 codebase is stated to be ready for volume distribution, our library is still given out as experimental. The old CC310 codebase was not meant for production.

### CODE REVIEW

**Comment by KaSroka** [comment]
@LuDuda @e-rk Please review.

File: examples/platforms/nrf528xx/nrf52840/README.md [hunk 1]
```diff
@@ -126,15 +126,107 @@ You can prefix the compiler command using the CCPREFIX parameter. This speeds up
...
+
+CryptoCell 310 platform implements abort mechanism. By default any fault in CryptoCell 310 library results in system reset. This behaviour can be changed by a call to following function which MUST be called before `mbedtls_platform_setup`:
+```
+void nrf_cc310_platform_set_abort(
+	nrf_cc310_platform_abort_apis_t const * const apis)
+```
+
+### CryptoCell 310 RTOS support
+
+Hardware accelerated mbedTLS library supports access from multiple threads. By default simple blocking implementation of mutexes is used. Used locking mechanism can be replaced using following function which MUST be called before `mbedtls_platform_setup`:
```
Comment (greg-fer) [inline examples/platforms/nrf528xx/nrf52840/README.md]:
```suggestion
The hardware-accelerated mbedTLS library supports access from multiple threads. By default, simple blocking implementation of mutexes is used. The locking mechanism can be replaced by using the `nrf_cc310_platform_set_mutexes()` function. This function must be called before `mbedtls_platform_setup`:
```

File: examples/platforms/nrf528xx/nrf52840/README.md [hunk 2]
```diff
@@ -126,15 +126,107 @@ You can prefix the compiler command using the CCPREFIX parameter. This speeds up
...
+```
+void nrf_cc310_platform_set_mutexes(nrf_cc310_platform_mutex_apis_t const * const apis,
+                                    nrf_cc310_platform_mutexes_t const * const mutexes)
+```
+
+Sample implementation of mutexes for FreeRTOS and Zephyr operating systems can be found in `nrf_cc310_platform_mutex_freertos.c` and `nrf_cc310_platform_mutex_zephyr.c`. Both files implements function `int nrf_cc310_platform_mutex_init(void)` which MUST be called before `mbedtls_platform_setup`.
```
Comment (greg-fer) [inline examples/platforms/nrf528xx/nrf52840/README.md]:
```suggestion
The sample implementation of mutexes for FreeRTOS and Zephyr operating systems can be found in `nrf_cc310_platform_mutex_freertos.c` and `nrf_cc310_platform_mutex_zephyr.c`. Both files implement the function `int nrf_cc310_platform_mutex_init(void)`. This function must be called before `mbedtls_platform_setup`.

File: examples/platforms/nrf528xx/nrf52840/README.md [hunk 3]
```diff
@@ -126,15 +126,107 @@ You can prefix the compiler command using the CCPREFIX parameter. This speeds up
...
+
+Sample implementation of abort functionality for FreeRTOS and Zephyr operating systems can be found in `nrf_cc310_platform_abort_freertos.c` and `nrf_cc310_platform_abort_zephyr.c`. Both files implements function `void nrf_cc310_platform_abort_init(void)` which MUST be called before `mbedtls_platform_setup`.
```
Comment (greg-fer) [inline examples/platforms/nrf528xx/nrf52840/README.md]:
```suggestion
The sample implementation of the abort functionality for FreeRTOS and Zephyr operating systems can be found in `nrf_cc310_platform_abort_freertos.c` and `nrf_cc310_platform_abort_zephyr.c`. Both files implement the function `void nrf_cc310_platform_abort_init(void)`. This function must be called before `mbedtls_platform_setup`.
```

File: examples/platforms/nrf528xx/nrf52840/README.md [hunk 4]
```diff
@@ -126,15 +126,107 @@ You can prefix the compiler command using the CCPREFIX parameter. This speeds up
...
+
+Typical setup in RTOS environment (error checking omitted):
```
Comment (greg-fer) [inline examples/platforms/nrf528xx/nrf52840/README.md]:
```suggestion
The typical setup in the RTOS environment (error checking omitted) is as follows:
```

File: examples/platforms/nrf528xx/nrf52840/README.md [hunk 5]
```diff
@@ -126,15 +126,107 @@ You can prefix the compiler command using the CCPREFIX parameter. This speeds up
...
+```
+mbedtls_platform_set_calloc_free(calloc, free);
+nrf_cc310_platform_abort_init();
+nrf_cc310_platform_mutex_init();
+mbedtls_platform_setup(NULL);
+```
+
+When building external application that uses RTOS with OpenThread libraries and CryptoCell 310 hardware acceleration please use:
```
Comment (greg-fer) [inline examples/platforms/nrf528xx/nrf52840/README.md]:
```suggestion
When building an external application that uses RTOS with OpenThread libraries and the CryptoCell 310 hardware acceleration, use the following configuration:
```
Shouldn't this be in a separate section?

File: examples/platforms/nrf528xx/nrf52840/README.md [hunk 6]
```diff
@@ -126,15 +126,107 @@ You can prefix the compiler command using the CCPREFIX parameter. This speeds up
...
+- Include directories:
+  - `third_party/NordicSemiconductor/libraries/nrf_security/mbedtls_platform_config`
+  - `third_party/NordicSemiconductor/libraries/nrf_security/include`
+  - `third_party/NordicSemiconductor/libraries/nrf_security/config`
+  - `third_party/NordicSemiconductor/libraries/nrf_security/nrf_cc310_platform/include`
+- Defines:
+  - `MBEDTLS_CONFIG_FILE` set to  `"nrf-config.h"`
+  - `MBEDTLS_USER_CONFIG_FILE` set to `"nrf52840-mbedtls-config.h"`
+- Sources:
+  - `nrf_cc310_platform_mutex_freertos.c` or `nrf_cc310_platform_mutex_zephyr.c` from `third_party/NordicSemiconductor/libraries/nrf_security/nrf_cc310_platform/src` depending on used RTOS
```
Comment (greg-fer) [inline examples/platforms/nrf528xx/nrf52840/README.md]:
```suggestion
  - `nrf_cc310_platform_mutex_freertos.c` or `nrf_cc310_platform_mutex_zephyr.c` from `third_party/NordicSemiconductor/libraries/nrf_security/nrf_cc310_platform/src`, depending on the RTOS used.
```

File: examples/platforms/nrf528xx/nrf52840/README.md [hunk 7]
```diff
@@ -126,15 +126,107 @@ You can prefix the compiler command using the CCPREFIX parameter. This speeds up
...
+  - `nrf_cc310_platform_abort_freertos.c` or `nrf_cc310_platform_abort_zephyr.c` from `third_party/NordicSemiconductor/libraries/nrf_security/nrf_cc310_platform/src` depending on used RTOS
```
Comment (greg-fer) [inline examples/platforms/nrf528xx/nrf52840/README.md]:
```suggestion
  - `nrf_cc310_platform_abort_freertos.c` or `nrf_cc310_platform_abort_zephyr.c` from `third_party/NordicSemiconductor/libraries/nrf_security/nrf_cc310_platform/src`, depending on the RTOS used.
```

File: examples/platforms/nrf528xx/nrf52840/README.md [hunk 8]
```diff
@@ -126,15 +126,107 @@ You can prefix the compiler command using the CCPREFIX parameter. This speeds up
...
+
+See [mbedTls Thread Safety and Multi Threading][mbedtls-thread-safety-and-multi-threading] for more details.
+
+### Optional software only mbedTLS threading support
+
+By default, software only mbedTLS library is built without support for multiple threads. You can enable this built-in support by building OpenThread with the following parameter:
```
Comment (greg-fer) [inline examples/platforms/nrf528xx/nrf52840/README.md]:
```suggestion
By default, the software-only mbedTLS library is built without support for multiple threads. You can enable this built-in support by building OpenThread with the following parameter:
```

File: examples/platforms/nrf528xx/nrf52840/README.md [hunk 9]
```diff
@@ -152,9 +244,21 @@ However, you can modify it, by providing a path to a header file with proper def
 $ make -f examples/Makefile-nrf52840 MBEDTLS_THREADING=1 MBEDTLS_THREADING_MUTEX_DEF="path_to_a_header_file_with_mutex_definition.h"
 ```
 
-See [mbedTls Thread Safety and Multi Threading][mbedtls-thread-safety-and-multi-threading] for more details.
+When building external application that uses RTOS with OpenThread libraries and software cryptography please use:
```
Comment (greg-fer) [inline examples/platforms/nrf528xx/nrf52840/README.md]:
```suggestion
When building an external application that uses RTOS with OpenThread libraries and software cryptography, use the following configuration:
```

File: examples/platforms/nrf528xx/nrf52840/README.md (hunk 5 above)
Comment (KaSroka) [reply examples/platforms/nrf528xx/nrf52840/README.md]: Well, it's in CC310 RTOS section and tells the configuration that you should use only for CC310 + RTOS so I think that this section is ok.

**Comment by KaSroka** [comment]
> Great work Kamil!
> 
> Can you please write a few words for community about the reason of this change, so someone can find reference later when this is merged.

@LuDuda You want it here or as a part of the commit message?

**Comment by LuDuda** [comment]
Either way is fine!

**Comment by KaSroka** [comment]
@LuDuda  I've done both :wink:

**Comment by KaSroka** [comment]
Rebased and ready for merge

**Comment by e-rk** [comment]
There is one problem that is more related to the exported heap api.

On IAR, [this line](https://github.com/KaSroka/openthread/blob/748531d3020ff3747d20cab82d884859950e3470/src/core/api/heap_api.cpp#L52) is unreachable because of the assert above.

**Comment by LuDuda** [comment]
@e-rk i faced this problem in the past. Please use `OT_UNREACHABLE_CODE` macro like [here](https://github.com/openthread/openthread/blob/master/src/ncp/ncp_base.cpp#L2150).

**Comment by jwhui** [comment]
> @e-rk i faced this problem in the past. Please use `OT_UNREACHABLE_CODE` macro like [here](https://github.com/openthread/openthread/blob/master/src/ncp/ncp_base.cpp#L2150).

Is this something we want to address in this PR?

**Comment by LuDuda** [comment]
@jwhui Yes, we will address it in this PR.

**Comment by KaSroka** [comment]
@jwhui Please don't merge this yet - we are waiting for nrf_security 0.9.1

**Comment by LuDuda** [comment]
Oops, I believe this PR was not ready for merge as Kamil stated above - @KaSroka?

@jwhui would it be possible to give us permission to set labels in PRs (e.g. work in progress)? Else I think we can try to change PR title to [DNM][.. in the future (DNM - do not merge).

**Comment by jwhui** [comment]
@LuDuda, apologies for that. I rolled master back. Please submit this PR and #4302 again.

**Pull request merged**